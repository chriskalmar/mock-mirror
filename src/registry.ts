import { DEFAULT_SCOPE } from './const';
import { logger } from './logger';

type MockedRoute = {
  pathPattern: string;
  method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH' | 'OPTIONS' | 'HEAD' | 'CONNECT' | 'TRACE' | 'ALL';
  response: unknown;
  status: number;
  headers?: Record<string, string>;
  contentType?: string;
  delay?: number;
};

type MockedRoutes = MockedRoute[];

const registry = new Map<string, MockedRoutes>();

registry.set(DEFAULT_SCOPE, []);

export const resetRegistry = () => {
  registry.clear();
  registry.set(DEFAULT_SCOPE, []);
};

export const addMockedRoute = ({ scope, route }: { scope: string; route: MockedRoute }) => {
  const routes = registry.get(scope) || [];

  registry.set(scope, [route, ...routes]);

  logger.info(`Added new route to scope ${scope}: ${route.method} ${route.pathPattern}`);
};

export const addMockedRoutes = ({ scope, routes }: { scope: string; routes: MockedRoute[] }) => {
  routes.forEach((route) => addMockedRoute({ scope, route }));
};

export const addDefaultMockedRoutes = (routes: MockedRoute[]) => {
  addMockedRoutes({ scope: DEFAULT_SCOPE, routes });
};

const findMatch = ({
  scope,
  path,
  method,
}: {
  scope: string;
  path: string;
  method: string;
}):
  | {
      route: MockedRoute;
      matchType: 'exact' | 'pattern';
    }
  | undefined => {
  const routes = registry.get(scope) ?? [];

  const exactMatchingRoute = routes.find((route) => route.pathPattern === path && route.method === method);

  if (exactMatchingRoute) {
    logger.info(
      `Found matching route in scope ${scope} for: ${method} ${path} [exact: ${exactMatchingRoute.pathPattern}]`,
    );

    return {
      route: exactMatchingRoute,
      matchType: 'exact',
    };
  }

  const exactMatchingFallbackRoute = routes.find((route) => route.pathPattern === path && route.method === 'ALL');

  if (exactMatchingFallbackRoute) {
    logger.info(
      `Found fallback route in scope ${scope} for: ALL ${path} [exact: ${exactMatchingFallbackRoute.pathPattern}]`,
    );

    return {
      route: exactMatchingFallbackRoute,
      matchType: 'exact',
    };
  }

  const matchingRoute = routes.find((route) => new RegExp(route.pathPattern).test(path) && route.method === method);

  if (matchingRoute) {
    logger.info(
      `Found matching route in scope ${scope} for: ${method} ${path} [pattern: ${matchingRoute.pathPattern}]`,
    );

    return {
      route: matchingRoute,
      matchType: 'pattern',
    };
  }

  if (method !== 'ALL') {
    const fallbackRoute = routes.find((route) => new RegExp(route.pathPattern).test(path) && route.method === 'ALL');

    if (fallbackRoute) {
      logger.info(`Found fallback route in scope ${scope} for: ALL ${path} [pattern: ${fallbackRoute.pathPattern}]`);

      return {
        route: fallbackRoute,
        matchType: 'pattern',
      };
    }
  }

  logger.warn(`No matching route found in scope ${scope} for: ${method} ${path}`);
};

export const findMatchingRoute = ({ scope, path, method }: { scope: string; path: string; method: string }) => {
  if (registry.has(scope)) {
    return findMatch({ scope, path, method });
  }

  logger.warn(`Scope ${scope} not found. Falling back to default scope: ${DEFAULT_SCOPE}`);

  return findMatch({ scope: DEFAULT_SCOPE, path, method });
};
