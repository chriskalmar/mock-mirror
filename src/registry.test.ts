import { beforeEach, describe, expect, it } from 'bun:test';
import { addDefaultMockedRoutes, addMockedRoutes, findMatchingRoute, resetRegistry } from './registry';

const scope = 'ff71ae5d-3112-4321-a9e1-459f72a453c0';

describe('registry', () => {
  beforeEach(() => {
    resetRegistry();
  });

  it('should find a route via exact matching', async () => {
    addMockedRoutes({
      scope,
      routes: [
        {
          pathPattern: '/',
          method: 'GET',
          response: 'hi',
          status: 200,
        },
        {
          pathPattern: '/lorem/ipsum',
          method: 'POST',
          response: 'dolor sit amet',
          status: 201,
        },
      ],
    });

    {
      const route = findMatchingRoute({ scope, path: '/', method: 'GET' });
      expect(route).toMatchSnapshot('/');
    }

    {
      const route = findMatchingRoute({ scope, path: '/lorem/ipsum', method: 'POST' });
      expect(route).toMatchSnapshot('/lorem/ipsum');
    }
  });

  it('should fall back to default scope', async () => {
    addDefaultMockedRoutes([
      {
        pathPattern: '/fallback',
        method: 'GET',
        response: 'fallback',
        status: 200,
      },
    ]);

    const route = findMatchingRoute({ scope: 'does-not-exist', path: '/fallback', method: 'GET' });
    expect(route).toMatchSnapshot('/fallback');
  });

  it('should fall back to ALL method', async () => {
    addMockedRoutes({
      scope,
      routes: [
        {
          pathPattern: '/catch-all',
          method: 'ALL',
          response: 'I got you covered!',
          status: 200,
        },
      ],
    });

    const route = findMatchingRoute({ scope, path: '/catch-all', method: 'GET' });
    expect(route).toMatchSnapshot('/catch-all');
  });

  it('should fall back to ALL method of default scope', async () => {
    addDefaultMockedRoutes([
      {
        pathPattern: '/catch-all',
        method: 'ALL',
        response: 'I got you covered!',
        status: 200,
      },
    ]);

    const route = findMatchingRoute({ scope: 'does-not-exist', path: '/catch-all', method: 'GET' });
    expect(route).toMatchSnapshot('/catch-all');
  });

  it('should find a route via pattern matching', async () => {
    addMockedRoutes({
      scope,
      routes: [
        {
          pathPattern: '/users/*',
          method: 'GET',
          response: 'any user',
          status: 200,
        },
        {
          pathPattern: '/users/12345',
          method: 'GET',
          response: 'user 12345',
          status: 200,
        },
        {
          pathPattern: '/api/**',
          method: 'GET',
          response: 'any api endpoint',
          status: 200,
        },
        {
          pathPattern: '/api/v1/**',
          method: 'GET',
          response: 'any v1 api endpoint',
          status: 200,
        },
        {
          pathPattern: '/api/v1/users/*',
          method: 'GET',
          response: 'any v1 user',
          status: 200,
        },
      ],
    });

    {
      const route = findMatchingRoute({ scope, path: '/users', method: 'GET' });
      expect(route).toBeUndefined();
    }

    {
      const route = findMatchingRoute({ scope, path: '/users/', method: 'GET' });
      expect(route).toBeUndefined();
    }

    {
      const route = findMatchingRoute({ scope, path: '/users/777', method: 'GET' });
      expect(route).toMatchSnapshot('/users/777');
    }

    {
      const route = findMatchingRoute({ scope, path: '/users/12345', method: 'GET' });
      expect(route).toMatchSnapshot('/users/12345');
    }

    {
      const route = findMatchingRoute({ scope, path: '/api', method: 'GET' });
      expect(route).toBeUndefined();
    }

    {
      const route = findMatchingRoute({ scope, path: '/api/', method: 'GET' });
      expect(route).toMatchSnapshot('/api/');
    }

    {
      const route = findMatchingRoute({ scope, path: '/api/lorem', method: 'GET' });
      expect(route).toMatchSnapshot('/api/lorem');
    }

    {
      const route = findMatchingRoute({ scope, path: '/api/v1', method: 'GET' });
      expect(route).toMatchSnapshot('/api/v1');
    }

    {
      const route = findMatchingRoute({ scope, path: '/api/v1/', method: 'GET' });
      expect(route).toMatchSnapshot('/api/v1/');
    }

    {
      const route = findMatchingRoute({ scope, path: '/api/v1/users', method: 'GET' });
      expect(route).toMatchSnapshot('/api/v1/users');
    }

    {
      const route = findMatchingRoute({ scope, path: '/api/v1/users/', method: 'GET' });
      expect(route).toMatchSnapshot('/api/v1/users/');
    }

    {
      const route = findMatchingRoute({ scope, path: '/api/v1/users/777', method: 'GET' });
      expect(route).toMatchSnapshot('/api/v1/users/777');
    }
  });

  it('should find a route via pattern matching with ALL method', async () => {
    addMockedRoutes({
      scope,
      routes: [
        {
          pathPattern: '/api/**',
          method: 'ALL',
          response: 'any api endpoint',
          status: 200,
        },
      ],
    });

    const route = findMatchingRoute({ scope, path: '/api/v1/users/777', method: 'POST' });
    expect(route).toMatchSnapshot('/api/v1/users/777');
  });

  it('should find a route via pattern matching with ALL method of default scope', async () => {
    addDefaultMockedRoutes([
      {
        pathPattern: '/api/**',
        method: 'ALL',
        response: 'any api endpoint',
        status: 200,
      },
    ]);

    const route = findMatchingRoute({ scope: 'does-not-exist', path: '/api/v1/users/777', method: 'POST' });
    expect(route).toMatchSnapshot('/api/v1/users/777');
  });

  it('should respect the order of routes', async () => {
    {
      resetRegistry();

      addMockedRoutes({
        scope,
        routes: [
          {
            pathPattern: '/api/users/*',
            method: 'GET',
            response: 'any user',
            status: 200,
          },
          {
            pathPattern: '/api/*',
            method: 'GET',
            response: 'any api endpoint',
            status: 200,
          },
        ],
      });

      const route = findMatchingRoute({ scope, path: '/api/users/777', method: 'GET' });
      expect(route).toMatchSnapshot('simple patterns');
    }

    {
      resetRegistry();

      addMockedRoutes({
        scope,
        routes: [
          {
            pathPattern: '/api/users/**',
            method: 'GET',
            response: 'any user',
            status: 200,
          },
          {
            pathPattern: '/api/**',
            method: 'GET',
            response: 'any api endpoint',
            status: 200,
          },
        ],
      });

      const route = findMatchingRoute({ scope, path: '/api/users/777', method: 'GET' });
      expect(route).toMatchSnapshot('wider patterns');
    }

    {
      resetRegistry();

      addMockedRoutes({
        scope,
        routes: [
          {
            pathPattern: '/api/**',
            method: 'GET',
            response: 'any api endpoint',
            status: 200,
          },
          {
            pathPattern: '/api/users/**',
            method: 'GET',
            response: 'any user',
            status: 200,
          },
        ],
      });

      const route = findMatchingRoute({ scope, path: '/api/users/777', method: 'GET' });
      expect(route).toMatchSnapshot('last pattern wins');
    }
  });
});
