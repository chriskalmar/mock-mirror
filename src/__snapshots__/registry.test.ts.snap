// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`registry should find a route via exact matching: / 1`] = `
{
  "matchType": "exact",
  "pathPattern": "/",
  "route": {
    "method": "GET",
    "pathPattern": "/",
    "response": "hi",
    "status": 200,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should find a route via exact matching: /lorem/ipsum 1`] = `
{
  "matchType": "exact",
  "pathPattern": "/lorem/ipsum",
  "route": {
    "method": "POST",
    "pathPattern": "/lorem/ipsum",
    "response": "dolor sit amet",
    "status": 201,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should fall back to default scope: /fallback 1`] = `
{
  "matchType": "exact",
  "pathPattern": "/fallback",
  "route": {
    "method": "GET",
    "pathPattern": "/fallback",
    "response": "fallback",
    "status": 200,
  },
  "scope": "[DEFAULT]",
}
`;

exports[`registry should fall back to ALL method: /catch-all 1`] = `
{
  "matchType": "exact",
  "pathPattern": "/catch-all",
  "route": {
    "method": "ALL",
    "pathPattern": "/catch-all",
    "response": "I got you covered!",
    "status": 200,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should fall back to ALL method of default scope: /catch-all 1`] = `
{
  "matchType": "exact",
  "pathPattern": "/catch-all",
  "route": {
    "method": "ALL",
    "pathPattern": "/catch-all",
    "response": "I got you covered!",
    "status": 200,
  },
  "scope": "[DEFAULT]",
}
`;

exports[`registry should find a route via pattern matching: /users/777 1`] = `
{
  "matchType": "pattern",
  "pathPattern": "/users/*",
  "route": {
    "method": "GET",
    "pathPattern": "/users/*",
    "response": "any user",
    "status": 200,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should find a route via pattern matching: /users/12345 1`] = `
{
  "matchType": "exact",
  "pathPattern": "/users/12345",
  "route": {
    "method": "GET",
    "pathPattern": "/users/12345",
    "response": "user 12345",
    "status": 200,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should find a route via pattern matching: /api/ 1`] = `
{
  "matchType": "pattern",
  "pathPattern": "/api/**",
  "route": {
    "method": "GET",
    "pathPattern": "/api/**",
    "response": "any api endpoint",
    "status": 200,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should find a route via pattern matching: /api/lorem 1`] = `
{
  "matchType": "pattern",
  "pathPattern": "/api/**",
  "route": {
    "method": "GET",
    "pathPattern": "/api/**",
    "response": "any api endpoint",
    "status": 200,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should find a route via pattern matching: /api/v1 1`] = `
{
  "matchType": "pattern",
  "pathPattern": "/api/**",
  "route": {
    "method": "GET",
    "pathPattern": "/api/**",
    "response": "any api endpoint",
    "status": 200,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should find a route via pattern matching: /api/v1/ 1`] = `
{
  "matchType": "pattern",
  "pathPattern": "/api/v1/**",
  "route": {
    "method": "GET",
    "pathPattern": "/api/v1/**",
    "response": "any v1 api endpoint",
    "status": 200,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should find a route via pattern matching: /api/v1/users 1`] = `
{
  "matchType": "pattern",
  "pathPattern": "/api/v1/**",
  "route": {
    "method": "GET",
    "pathPattern": "/api/v1/**",
    "response": "any v1 api endpoint",
    "status": 200,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should find a route via pattern matching: /api/v1/users/ 1`] = `
{
  "matchType": "pattern",
  "pathPattern": "/api/v1/**",
  "route": {
    "method": "GET",
    "pathPattern": "/api/v1/**",
    "response": "any v1 api endpoint",
    "status": 200,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should find a route via pattern matching: /api/v1/users/777 1`] = `
{
  "matchType": "pattern",
  "pathPattern": "/api/v1/users/*",
  "route": {
    "method": "GET",
    "pathPattern": "/api/v1/users/*",
    "response": "any v1 user",
    "status": 200,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should find a route via pattern matching with ALL method: /api/v1/users/777 1`] = `
{
  "matchType": "pattern",
  "pathPattern": "/api/**",
  "route": {
    "method": "ALL",
    "pathPattern": "/api/**",
    "response": "any api endpoint",
    "status": 200,
  },
  "scope": "ff71ae5d-3112-4321-a9e1-459f72a453c0",
}
`;

exports[`registry should find a route via pattern matching with ALL method of default scope: /api/v1/users/777 1`] = `
{
  "matchType": "pattern",
  "pathPattern": "/api/**",
  "route": {
    "method": "ALL",
    "pathPattern": "/api/**",
    "response": "any api endpoint",
    "status": 200,
  },
  "scope": "[DEFAULT]",
}
`;
